odoo/orm/domains.py

```diff
@@ -897,6 +935,51 @@ def _optimize_field_search_method(self, model: BaseModel) -> Domain:
             model_label=f"{model.env['ir.model']._get(model._name).name!r} ({model._name})",
         ))
 
+    def filter_records(self, records: M) -> M:
+        if not records:
+            return records
+        if self._opt_level < OptimizationLevel.BASIC:
+            return self.optimize(records, full=False).filter_records(records)
+        operator = self.operator
+        if operator not in STANDARD_CONDITION_OPERATORS:
+            # basic optimization was done before calling this, we may resort to SQL
+            return self.optimize(records, full=True).filter_records(records)
+        field_expr, value = self.field_expr, self.value
+        positive_operator = NEGATIVE_CONDITION_OPERATORS.get(operator, operator)
+
+        if isinstance(value, SQL):
+            # transform into an Query value
+            if positive_operator == operator:
+                condition = self
+                operator = 'any'
+            else:
+                condition = ~self
+                operator = 'not any'
+            positive_operator = 'any'
+            field_expr = 'id'
+            value = records.with_context(active_test=False)._search(DomainCondition('id', 'in', OrderedSet(records.ids)) & condition)
+            assert isinstance(value, Query)
+        if isinstance(value, Query):
+            # rebuild a domain with an 'in' values
+            if positive_operator not in ('in', 'any'):
+                self._raise("Cannot filter using Query without the 'any' or 'in' operator")
+            if positive_operator == 'any':
+                operator = 'in' if positive_operator == operator else 'not in'
+                positive_operator = 'in'
+            value = set(value.get_result_ids())
+            return DomainCondition(field_expr, operator, value).filter_records(records)
+
+        field = self._field(records)
+        if field_expr == 'display_name':
+            # when searching by name, ignore AccessError
+            field_expr = 'display_name.no_error'
+        elif field_expr == 'id':
+            # for new records, compare to their origin
+            field_expr = 'id.origin'
+        func = field.filter_function(records, field_expr, positive_operator, value)
+        final_func = func if positive_operator == operator else lambda rec: not func(rec)
+        return records.browse(rec_id for rec_id, rec in zip(records._ids, records) if final_func(rec))
```

Comment:
That creates a tons of records if your domain is big and most of the records matches the domain.
What about returning a _filter_function also for domain objects? :thinking:

Reply:
Adapted.